SL/Controller/ZUGFeRD.pm

```diff
@@ -77,72 +97,174 @@ sub action_import_zugferd {
 
-  my $iban = $dom->findnodes('//ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeSettlementPaymentMeans' .
-                             '/ram:PayeePartyCreditorFinancialAccount/ram:IBANID')->string_value;
-  my $ibanmessage;
+  # Create a record template for this imported invoice
+  $template_ap = SL::DB::RecordTemplate->new(
+      vendor_id=>$vendor->id,
+  );
+
+  # Check IBAN specified on bill matches the one we've got in
+  # the database for this vendor.
   $ibanmessage = $iban ne $vendor->iban ? "Record IBAN $iban doesn't match vendor IBAN " . $vendor->iban : $iban if $iban;
 
+  # Use invoice creation date as due date if there's no due date
+  $metadata{'duedate'} = $metadata{'transdate'} unless defined $metadata{'duedate'};
+
+  # parse dates to kivi if set/valid
+  foreach my $key ( qw(transdate duedate) ) {
+    next unless defined $metadata{$key};
+    $metadata{$key} =~ s/^\s+|\s+$//g;
+
+    if ($metadata{$key} =~ /^([0-9]{4})-?([0-9]{2})-?([0-9]{2})$/) {
+    $metadata{$key} = DateTime->new(year  => $1,
+                                    month => $2,
+                                    day   => $3)->to_kivitendo;
+    }
+  }
+
+  # Try to fill in AP account to book against
+  my $ap_chart_id = $::instance_conf->get_ap_chart_id;
+
+  unless ( defined $ap_chart_id ) {
+    # If no default account is configured, just use the first AP account found.
+    my $ap_chart = SL::DB::Manager::Chart->get_all(
+      where   => [ link => 'AP' ],
+      sort_by => [ 'accno' ],
+    );
+    $ap_chart_id = ${$ap_chart}[0]->id;
+  }
+
+  my $currency = SL::DB::Manager::Currency->find_by(
+    name => $metadata{'currency'},
+    );
+
+  $template_ap->assign_attributes(
+    template_name       => "Faktur-X/ZUGFeRD/XRechnung Import $vendor->name, $invnumber",
+    template_type       => 'ap_transaction',
+    direct_debit        => $metadata{'direct_debit'},
+    notes               => "Faktur-X/ZUGFeRD/XRechnung Import. Type: $metadata{'type'}\nIBAN: " . $ibanmessage,
+    taxincluded         => 0,
+    currency_id         => $currency->id,
+    ar_ap_chart_id      => $ap_chart_id,
+    );
+
+  $template_ap->save;
+
+  my $default_ap_amount_chart = SL::DB::Manager::Chart->find_by(charttype => 'A');
+
+  foreach my $i ( @items )
+    {
+    my %item = %{$i};
+
+    my $net_total = $item{'subtotal'};
+    my $desc = $item{'description'};
+    my $tax_rate = $item{'tax_rate'} / 100; # XML data is usually in percent
+
+    my $taxes = SL::DB::Manager::Tax->get_all(
+      where   => [ chart_categories => { like => '%' . $default_ap_amount_chart->category . '%' },
+                   rate => $tax_rate,
+                 ],
+    );
+
+    # If we really can't find any tax definition (a simple rounding error may
+    # be sufficient for that to happen), grab the first tax fitting the default
+    # category, just like the AP form would do it for manual entry.
+    if ( scalar @{$taxes} == 0 ) {
+      $taxes = SL::D::ManagerTax->get_all(
```

Comment:
Hier gibt's ne Fehlermeldung wegen eines sehr offensichtlichen Schreibfehlers:

```
Can't locate object method "get_all" via package "SL::D::ManagerTax" (perhaps you forgot to load "SL::D::ManagerTax"?)
```

Du meinst bestimmt `SL::DB::Manager::Tax`.

Reply:
In der Tat. Repariert.